dotnet/src/SemanticKernel.Core/Data/TextSearchBehavior/TextSearchBehavior.cs

```diff
@@ -0,0 +1,148 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Text;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.AI;
+
+namespace Microsoft.SemanticKernel.Data;
+
+/// <summary>
+/// A component that does a search based on any messages that the AI model is invoked with and injects the results into the AI model invocation context.
+/// </summary>
+[Experimental("SKEXP0130")]
+public sealed class TextSearchBehavior : AIContextBehavior
+{
+    private const string DefaultPluginSearchFunctionName = "Search";
+    private const string DefaultPluginSearchFunctionDescription = "Allows searching for additional information to help answer the user question.";
+    private const string DefaultContextPrompt = "## Additional Context\nConsider the following information from source documents when responding to the user:";
+    private const string DefaultIncludeCitationsPrompt = "Include citations to the source document with document name and link if document name and link is available.";
+
+    private readonly ITextSearch _textSearch;
+
+    private readonly AIFunction[] _aIFunctions;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="TextSearchBehavior"/> class.
+    /// </summary>
+    /// <param name="textSearch">The text search component to retrieve results from.</param>
+    /// <param name="options">Options that configure the behavior of the component.</param>
+    /// <exception cref="ArgumentNullException"></exception>
+    public TextSearchBehavior(ITextSearch textSearch, TextSearchBehaviorOptions? options = default)
+    {
+        Verify.NotNull(textSearch);
+
+        this._textSearch = textSearch;
+        this.Options = options ?? new();
+
+        this._aIFunctions =
+        [
+            AIFunctionFactory.Create(
+            this.SearchAsync,
+            name: this.Options.PluginFunctionName ?? DefaultPluginSearchFunctionName,
+            description: this.Options.PluginFunctionDescription ?? DefaultPluginSearchFunctionDescription)
+        ];
+    }
+
+    /// <summary>
+    /// Gets the options that have been configured for this component.
+    /// </summary>
+    public TextSearchBehaviorOptions Options { get; }
+
+    /// <inheritdoc/>
+    public override async Task<AIContextPart> OnModelInvokeAsync(ICollection<ChatMessage> newMessages, CancellationToken cancellationToken = default)
+    {
+        if (this.Options.SearchTime != TextSearchBehaviorOptions.RagBehavior.BeforeAIInvoke)
+        {
+            return new()
+            {
+                AIFunctions = this._aIFunctions.ToList(),
```

Comment:
Do `AIFunctions` need to be a list?

```suggestion
                AIFunctions = this._aIFunctions,
```